File: tuesday/morning/claude_exercise/figure3.py
import time

N_STEPS = 100000

def main():
    # Initial condition
    x0 = 0.3
    trajectory = [0.0] * (N_STEPS + 1)
    trajectory[0] = x0
    x_current = x0

    # Start timer
    start = time.time()

    # Generate long time series using quadratic map g(x) = 4x(1-x)
    for t in range(1, N_STEPS + 1):
        x_current = 4.0 * x_current * (1.0 - x_current)
        trajectory[t] = x_current

    # End timer
    elapsed_time = time.time() - start

    # Write to file
    with open('figure3_data_py.txt', 'w') as fp:
        fp.write('# x_t\n')
        for t in range(N_STEPS + 1):
            fp.write(f'{trajectory[t]:.10f}\n')

    print(f'Figure 3 data written to figure3_data_py.txt')
    print(f'Total observations: {N_STEPS + 1}')
    print(f'Execution time: {elapsed_time:.6f} seconds')

if __name__ == '__main__':
    main()

File: tuesday/morning/claude_exercise/figure3.f90
program figure3
    implicit none
    integer, parameter :: n_steps = 100000
    real(8) :: x0, x_current
    real(8), dimension(0:n_steps) :: trajectory
    integer :: t
    real(8) :: start_time, end_time, elapsed_time
    integer :: count_rate, count_max
    integer :: start_count, end_count

    ! Initial condition
    x0 = 0.3d0
    trajectory(0) = x0
    x_current = x0

    ! Start timer
    call system_clock(start_count, count_rate, count_max)

    ! Generate long time series using quadratic map g(x) = 4x(1-x)
    do t = 1, n_steps
        x_current = 4.0d0 * x_current * (1.0d0 - x_current)
        trajectory(t) = x_current
    end do

    ! End timer
    call system_clock(end_count)
    elapsed_time = real(end_count - start_count) / real(count_rate)

    ! Write to file
    open(unit=10, file='figure3_data.txt', status='replace')
    write(10, '(A)') '# x_t'
    do t = 0, n_steps
        write(10, '(F20.10)') trajectory(t)
    end do
    close(10)

    print *, 'Figure 3 data written to figure3_data.txt'
    print *, 'Total observations:', n_steps + 1
    print '(A,F10.6,A)', 'Execution time: ', elapsed_time, ' seconds'

end program figure3
